File: PA_21.py
#LC 256 - Paint House
#Time Complexity - O(n) Not sure, can you comment?
#Space Complexity - O(n)
class Solution(object):
    def minCost(self, costs):
        """
        :type costs: List[List[int]]
        :rtype: int
        """
        if len(costs) == 0:
            return 0
        
        if len(costs) == 1:
            return min(costs[0])
        
        def helper(costs,red,blue,green,index):
            #print(red,green,blue,index)
            if index == len(costs):
                #print(red,blue,green)
                return min(red,green,blue)
            tred = costs[index][0] + min(green,blue)
            tblue = costs[index][1] + min(green,red)
            tgreen = costs[index][2] + min(red,blue)
            #print(tred,tblue,tgreen)
            ans = helper(costs,tred,tblue,tgreen,index+1)
            return ans
        return helper(costs,costs[0][0],costs[0][1],costs[0][2],1)

File: PA21.java
//Leetcode 256. Paint House
//Time Complexity : O(n) //n is no of houses
//Space Compplexity : O(1)

class Solution {
    public int minCost(int[][] costs) {
        if(costs== null || costs.length==0 ){
            return 0;
        }
        int bl=costs[0][0];
        int r=costs[0][1];
        int g=costs[0][2];
        int tbl=0;
        int tr=0;
        int tg=0;
        for (int i=1; i<costs.length;i++){
          
                tbl=bl;
                tr=r;
                tg=g;
                bl= Math.min(tr,tg)+costs[i][0];
                r= Math.min(tbl,tg)+costs[i][1];
                g= Math.min(tr,tbl)+costs[i][2];
            
        }
        return Math.min(Math.min(r,g),bl);
    }
}